Demo Update - Allow User To Preview, Add And Edit CSV/TSV File Headers

scottcarol opened the pull request:
This pull request addresses #106 and adds functionality and testing to the Demo. The goal is to help new users successfully use the pivot table by letting them preview the headers of their uploaded files, edit those headers if desired, and add headers if their CSV/TSV file did not include them.

Specific functionality added:
- File upload now limited to CSV/TSV (previous file upload allowed all file extensions)
- If "Use Sample Dataset" is toggled and a file is uploaded, user sees previews of headers
- User can edit those uploaded headers by clicking "Edit Headers" button
- User can add a header row by clicking "Oops, I Need To Add My Headers!" button
- Validation does not allow a user to add an empty header row
- Unit testing added in `app.spec.js`

Notes on changes:
- In `package.json`, added Toggle-Display library for conditional display
- In `package.json`, added `.css` to the list of extensions mocked in unit testing
- In `package.json`, changed `enzyme-adapter-react-16' from "^1.1.0" to "1.1.0" because a newer version was causing errors in 'Drawer.spec.js'

scottcarol committed:
UI complete, now working on Jest testing

scottcarol committed:
Spec test setup complete

scottcarol committed:
Building out app.spec.js

scottcarol committed:
tests written and passing

scottcarol committed:
Restoring large csv file to sampledata

scottcarol committed:
restoring original webpack demo config settings

scottcarol committed:
restoring webpack demo config syntax

turnerniles commented:
@scottcarol this looks awesome and thank you! I will take a look at it this week and review / give feedback or merge in depending! I was away this weekend so I didn't have the chance.

scottcarol commented:
Awesome @turnerniles definitely open to feedback!

turnerniles commented on src/app.jsx:
```diff
@@ -242,7 +312,7 @@ export default class App extends React.Component {
             </div>
             <Select
-              name="Dataset"
+              name="ColorPack"
```
Oh thanks for changing this!

turnerniles commented on styles/index.scss:
```diff
@@ -73,7 +74,7 @@ section.react-virtualized-pivot-demo{
     display: flex;
     flex-wrap: wrap;
-    align-items: center;
+    //align-items: center;
```
Alignment looks good without `align-items: center;`. Let's delete the commented out line?

turnerniles commented on src/app.jsx:
```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
+      newHeaderRow[id] = value;
+      this.setState({headerRow: newHeaderRow});
+    } else {
+      throw Error('null value');
```
I see you need this else statement because you can't pivot the data when the headers are null values. But it causes a weird experience where you can't delete the last character in the input box.

<img width="518" alt="screen shot 2018-04-16 at 12 29 18 am" src="https://user-images.githubusercontent.com/8146241/38790179-40b9703a-410d-11e8-9596-f2f8e8fb4cbb.png">

I think we can reuse your create new validation logic to fix this:

```
  editHeader(evt) {
    const id = evt.target.id;
    const value = evt.target.value;
    const newHeaderRow = this.state.headerRow.slice();

    newHeaderRow[id] = value;

    this.setState({headerRow: newHeaderRow});
  }
```

```
  saveEditedHeaders() {
    const newDataSet = this.state.data.slice();

    let validHeaders = true;

    this.state.headerRow.forEach(header => {
      if (header === null || header === '') {
        validHeaders = false;
      }
    });
    if (validHeaders === false) {
      this.setState({validationMessage: 'Headers cannot be blank.'});
    } else {
      newDataSet[0] = this.state.headerRow;
      this.setState({data: newDataSet, showEditHeaderForm: false,
        validationMessage: null});
    }
  }
```

In fact it seems it will be the same function as addHeaders so we could just use the same function in both cases and delete one. What do you think?

turnerniles commented on src/app.jsx:
```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
+    let validHeaders = true;
+
+    this.state.headerRow.forEach(header => {
+      if (header === null || header === '') {
```
So the header will be null when input box is created but '' if you enter and then delete text. Nice!

turnerniles commented on src/app.jsx:
```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
+  addHeaders() {
+    const newDataSet = this.state.data.slice();
+
+    let validHeaders = true;
```
You could use the use the reduce patten instead to declare this variable. See below. I'll leave it to you if you want to change it or not though!

```
    const validHeaders = this.state.headerRow.reduce((acc, header) => {
      if (header === null || header === '') {
        acc = false;
      }
      return acc;
    }, true);
```

turnerniles commented on src/app.jsx:
```diff
@@ -205,32 +291,16 @@ export default class App extends React.Component {
         </div>
+        <div className='toggle-user-data'>
+          <input type="radio" value="true" id="default"
+            onChange={this.toggleUserData.bind(this)}
```
Change to ```this.toggleUserData``` because we bind this to toggleUserData in the constructor function ```    this.toggleUserData = this.toggleUserData.bind(this);```.

turnerniles commented on src/app.jsx:
```diff
@@ -205,32 +291,16 @@ export default class App extends React.Component {
+            defaultChecked name="dataUsed"/> Use Sample Dataset
+          <br/>
+          <input type="radio" value="false" id="custom"
+            onChange={this.toggleUserData.bind(this)}
```
Change to this.toggleUserData as per other comment.

turnerniles commented on src/app.jsx:
```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
+
+  editHeader(evt) {
+    const id = evt.target.id;
+
```
Nitpicky style I know. But could you delete the return between consts?

turnerniles commented:
@scottcarol This is super cool! Wow! It's great to have the option to edit the headers or add a new header. 👍. I left you some comments that I think should be fairly quick to change and then we can merge. Let me know if you have any comments/questions. I will be quicker to respond this week.

scottcarol committed:
Incorporating code review changes

scottcarol committed:
Deleting extra line break

scottcarol commented:
Great feedback @turnerniles. I combined `addHeaders` and `saveEditedHeaders` into a new function called `saveNewHeaders` and moved the form validation there as well. Test also updated.

turnerniles commented:
@scottcarol Amazing!! Thank you! Merging now.

scottcarol commented:
@turnerniles Happy to help, thanks for creating this great library.
